---
title: "1.1_datawrangle"
output: html_document
date: "2022-10-02"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Load libraries
```{r}
#load libraries 
library(dartR)
library(adegenet)
library(poppr)
library(hierfstat)
library(reshape2)
library(ggplot2)
library(RColorBrewer)
library(scales)
library(dplyr)
```

**This markdown goes through the initial filtering stages for Acropora humilis using Dartseq data** 


#Load data
```{r}
#we are loading the dart SNP file (first file) and the metadata that has the individuals and the population they were sampled from
gl <- gl.read.dart(filename = "Report_DAc22-6709_SNP_mapping_2.csv", covfilename = "Ahum_CSMP_LIMS.csv")

gl <- gl.drop.ind(gl, 
                    ind.list= c("A ten 271", "A ten 289"))
```


# Check metadata
```{r}
#double checking metadata, number of indivs, loci and populations 
nInd(gl)
#262 individuals

nPop(gl)
popNames(gl)
#we have 11 populations/reefs 

nLoc(gl)
#74335 loci in the original dataset (gl)

#converting SNP files to a matrix 
#generate a matrix of the SNP scores, with 0 as homozygous reference, 2 as homozygous alternate and 1 as heterozygous
m <- as.matrix(gl)

gl.report.callrate(gl, method = "ind")
#gl.filter.callrate(method = "ind")
```

#1. Filtering steps

##1.1 Filter secondaries
```{r}
#First we are filtering SNP call rate typically at 95 % threshold to make a confident call on genotype. 
#we need to decide whether it's better to filter by individual or by locus first. We agree it is better to lose loci in this case than to lose individuals, so we will filter by loci first. Indicated by method = "loc"
# We are basically calculating the percentage of complete genotypes per loci in the entire dataset for Acropora. 

#filter out secondaries. Dart will include fragements with more than one SNP but record separately with same Clone ID (= Allele ID). So we need to filter these out as multiple SNP loci within a fragment (secondaries) are likely to be linked. Can either choose to do by (method = 'best') or (method = 'random). Default is random. This is info from: 
#    https://rdrr.io/cran/dartR/man/gl.filter.secondaries.html

#filter secondaries. this helps with linkage disequilibrium 
gl <- gl.filter.secondaries(gl)
nLoc(gl)
#39,699 loci now
```

##1.2 Filter reproducibility 
```{r}
gl <- gl.filter.reproducibility(gl, threshold = 0.98)
nLoc(gl)
```

##1.3 FIlter call rate
```{r}
#filter loci with call rates less than 0.90 
gl2.80 <- gl.filter.callrate(gl, method = "loc", threshold = 0.80)
nLoc(gl2.80)
#8716 loci

gl2.85 <- gl.filter.callrate(gl, method = "loc", threshold = 0.85)
nLoc(gl2.85)
#6887 loci 

gl2.90 <- gl.filter.callrate(gl, method = "loc", threshold = 0.90)
nLoc(gl2.90)
#5045

```



##1.4 Filter read depth
```{r}
gl.report.rdepth(gl2.90)
#filter read depth < 5x
gl2.90 <- gl.filter.rdepth(gl2.80, lower = 5, upper = 200, verbose = 3)

nLoc(gl2.90)
#7514
```


##1.5 Filter MAFs
```{r}
#MAFs provide information about rare vs. common variants in a population. 
#Rare varients are < 0.05. #common variants are > 0.05 
#this will give an idea of the variation of genotypes for a given SNP in a population. If MAF for a population is low, it implies the major allele for the SNP is conserved and fixed, potentially. If MAF is high, it indicates the SNP is not conserved.

#If for a given SNP you have missing data in all populations but one, we would exclude such SNP from the analysis (but if you applied the 95% missingness filter you probably already settled this). 
#If you have a SNP that assumes a particular genotype only in one population, then keep it, as this SNP could be very informative about population structure and local adaptation


#We will filter out loci with a MAF less than 0.05.**
gl3 <- gl.filter.maf(gl2.90, threshold = 0.95) 

#save object so we don't have to run the above script each time 
#save(gl3, file = "gl3.RData")

#report for MAF after filtering with MAF < 0.05 
gl.report.maf(gl3)
nLoc(gl3)
#4214 snps remaining 

save(gl3, file = "gl3.RData")
```


##1.6 Impute missing data 
```{r, include=FALSE}
#here we are imputing any missing data using the 'neighbour' method. this is essential for some analyses later on to not have missing data. The advantage of the 'neighbour' approach (compared to 'HW' or 'frequency')is that it works regardless of how many individuals are in the population to which the focal individual belongs, and the displacement of the individual is haphazard as opposed to drawing individual toward the population centroid with HW or Freq. 
#Better for our analysis 
#a lot of downstream analysis cannot deal with missing data 

gl3 <- gl.impute(
  gl3,
  method = "neighbour",
  fill.residual = TRUE,
  parallel = FALSE,
  verbose = NULL
)
```

#2. Additional Filtering

##2.1 Convert to genind
```{r}
#converting a genlight object (we have) to a genind object so we can look at genetic data at an individual data using adegenet package. 
acro_gen <- gl2gi(gl3, probar = FALSE, verbose = NULL)
acro_gen

is.genind(acro_gen)
```


##2.2 Check relatedness
```{r}
#here we are checking for any clones / duplicate samples. We want to remove any that are. 
#check genotypes are unique
acro_gen_mlg <- mlg(acro_gen)

#we don't have any clones present in dataset
```

##2.3 Filter monomorphic loci
```{r}
#check that loci are still polymorphic after filtering. Retaining monomorphic loci unnecessarily increases the size of the dataset and will affect some calculations.

isPoly(acro_gen) %>% summary

#remove loci that are not polymorphic

poly_loci = names(which(isPoly(acro_gen) == TRUE))
acro_gen = acro_gen[loc = poly_loci]
isPoly(acro_gen) %>% summary
```

```{r}
save(acro_gen, file = "acro_gen.RData")
```


#3. Genetic diversity statistics

## 3.1 Heterozygosity (Ho)
```{r}
#Here we are calculating basic stats using hierfstat
basic_acro = basic.stats(acro_gen, diploid = TRUE)

# Mean observed heterozygosity per site
Ho_acro = apply(basic_acro$Ho, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)
Ho_acro
```

## 3.2 Heterozygosity (He)
```{r}
# Mean expected heterozygosity per site
He_acro = apply(basic_acro$Hs, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)
He_acro
```


## 3.3 Plot He/Ho
```{r}
# Create a data.frame of site names, Ho and He and then convert to long format
Het_acro_df = data.frame(Site = names(Ho_acro), Ho = Ho_acro, He = He_acro) %>%
  melt(id.vars = "Site")

# Custom theme for ggplot2
custom_theme = theme(
  axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, face = "bold"),
  axis.text.y = element_text(size = 10),
  axis.title.y = element_text(size = 12),
  axis.title.x = element_blank(),
  axis.line.y = element_line(size = 0.5),
  legend.title = element_blank(),
  legend.text = element_text(size = 12),
  panel.grid = element_blank(),
  panel.background = element_blank(),
  plot.title = element_text(hjust = 0.5, size = 15, face="bold")
  )

# Italic label
hetlab.o = expression(italic("H")[o])
hetlab.e = expression(italic("H")[e])

# Acropora heterozygosity barplot
ggplot(data = Het_acro_df, aes(x = Site, y = value, fill = variable))+
  geom_bar(stat = "identity", position = position_dodge(width = 0.6), colour = "black")+
  scale_y_continuous(expand = c(0,0), limits = c(0,0.2), breaks = c(0, 0.05, 0.10, 0.15, 0.20))+
  scale_fill_manual(values = c("pink", "#bdbdbd"), labels = c(hetlab.o, hetlab.e))+
  ylab("Heterozygosity")+
  ggtitle("Acropora")+
  custom_theme
```
**Ho referes to observed heterozygosity in pink. He refers to expected heterozygosity in grey.**
**Low observed heterozygosity means little genetic variability = low gene flow** 
**If heterozygosity is lower than expected, we attribute this to inbreeding, which makes sense in areas such as the Coral Sea with low gene flow** 


##3.4 Inbreeding (Fis)
```{r}
#Looking at Fis values. This is an inbreeding coefficient. Ranges from 0-1. A high FIS implies a lot of inbreeding in a population. 
#(e.g. Bougainville, Frederick, Wreck)

apply(basic_acro$Fis, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)
```


##3.5 Isolation by distance models
```{r}
ibd <- gl.ibd(gl3, Dgeo_trans='log(Dgeo)' ,Dgen_trans='Dgen/(1-Dgen)')
```


##3.6 Calculate Fst
```{r}
#using hierfstat package here. 
#compute pairwise Fst values. Returns pairwise distances folloiwng Weir & Cockerham 1984 
acro_fst = genet.dist(acro_gen, method = "WC84") %>% round(digits = 3)
acro_fst
```


##3.7 Visualise Fst
```{r}

#Visualise pairwise Fst for reefs 
# Desired order of labels

lab_order = c("Osprey", "Bougainville", "Moore", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick","Wreck")


# Change order of rows and cols
fst.mat = as.matrix(acro_fst)
fst.mat1 = fst.mat[lab_order, ]
fst.mat2 = fst.mat1[, lab_order]

# Create a data.frame
ind = which(upper.tri(fst.mat2), arr.ind = TRUE)
fst.df = data.frame(Site1 = dimnames(fst.mat2)[[2]][ind[,2]],
                    Site2 = dimnames(fst.mat2)[[1]][ind[,1]],
                    Fst = fst.mat2[ ind ])

# Keep the order of the levels in the data.frame for plotting 
fst.df$Site1 = factor(fst.df$Site1, levels = unique(fst.df$Site1))
fst.df$Site2 = factor(fst.df$Site2, levels = unique(fst.df$Site2))
fst.df$Fst[fst.df$Fst < 0] = 0
fst.df %>% str
fst.label = expression(italic("F")[ST])
mid = max(fst.df$Fst) / 2

# Plot heatmap
ggplot(data = fst.df, aes(x = Site1, y = Site2, fill = Fst)) +
  geom_tile(colour = "black") +
  # geom_text(aes(label = Fst), color="black", size = 3)+
  scale_fill_gradientn(
    colours = c("#66BD63", "#FFFFBF", "#D73027"),
    limits = c(0, .18),
    breaks = c(0, 0.09, 0.18),
    guide = "colorbar"
  ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0), position = "right") +
  theme(
    axis.text = element_text(colour = "black", size = 10, face = "bold"),
    axis.title = element_blank(),
    axis.text.x = element_text(angle = 45, vjust = 0.5),
    panel.grid = element_blank(),
    panel.background = element_blank(),
         legend.position = c(0.15, 0.7),  # Adjust the coordinates (x, y) here
    legend.title = element_text(size = 14, face = "bold"),
    legend.text = element_text(size = 10),
    legend.direction = "horizontal"
  ) +
  coord_fixed(ratio = 0.9)
```


##3.8 Allelic richness
```{r}
allelic.richness(genind2hierfstat(acro_gen))$Ar %>%
  apply(MARGIN = 2, FUN = mean) %>% 
  round(digits = 3)
```


##3.9 AMOVA
```{r}
#running Analysis of Molecular Variance
library("apex")
library("adegenet")
library("pegas")
library("mmod")
library("poppr")

# Import strata
my_strata <- read.csv("acro_strata.csv") %>% 
  dplyr::select(Ind, Pop) 

#define strata file in genind
strata(acro_gen) <- my_strata
setPop(acro_gen) <-~Pop
```


```{r}
# AMOVA
amova.sector <- poppr.amova(acro_gen , ~Pop,  within = T) 
amova.sector
# this is to use populations as lowest level

amova.sector_sig <- randtest(amova.sector, nrepet = 999)

# This was done with no correction method though!

amova.sector_sig
plot(amova.sector_sig)

#most of the variation occurs within samples, but we also have a lot of variation between samples. 
#while there is variation between sectors, there is not a lot, and suggests relatively panmictic populations. 
```



#4. PCA

##4.1 PCA model 
```{r}
x = tab(acro_gen, NA.method = "mean")

# Perform PCA
pca1 = dudi.pca(x, scannf = FALSE, scale = FALSE, nf = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent = pca1$eig/sum(pca1$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,12),
        names.arg = round(percent, 1))
```


##4.2 PCA coord dataframe 
```{r}
# Create a data.frame containing individual coordinates
ind_coords = as.data.frame(pca1$li) 

# Rename columns of dataframe
colnames(ind_coords) = c("Axis1","Axis2","Axis3")

# Add a column containing individuals
ind_coords$Ind = indNames(acro_gen)

# Add a column with the site IDs
ind_coords$Site = acro_gen$pop

# Calculate centroid (average) position for each population
centroid = aggregate(cbind(Axis1, Axis2, Axis3) ~ Site, data = ind_coords, FUN = mean)

# Add centroid coordinates to ind_coords dataframe
ind_coords = left_join(ind_coords, centroid, by = "Site", suffix = c("",".cen"))

# Define colour palette
cols = brewer.pal(nPop(acro_gen), "Set3")

# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

```


```{r}
#we need to extract Axis1 and Axis2 to serve at host genetic factors in our glmm 
ind.coords.glmm <- ind_coords %>% dplyr::select(Ind, Axis1, Axis2) %>% 
  rename(HostPC1 = Axis1) %>% 
  rename(HostPC2 = Axis2)

save(ind.coords.glmm, file = "ind.coords.glmm.RData")
```


##4.3 PCA metadata 
```{r}
load("meta_acro.RData")
meta_acro <- meta_acro %>% dplyr::select(Vial, catBleaching, Depth, DHW, Lat, Long, MMM, recent.maxDHW) %>% 
  rename(Ind = Vial)

ind_coords.full <- ind_coords %>% left_join(meta_acro)

save(ind_coords.full, file = "ind_coords.full.RData")
```





##4.4 Theme for PCA
```{r}
theme_PCA <-theme(aspect.ratio = 3/5) +
  theme(axis.line = element_blank()) +
  theme(plot.background = element_blank(),
    panel.background = element_blank(),
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank()) +
 theme(panel.border = element_rect(colour = "black", fill=NA, linewidth=0.5))
```


##4.5 Plot PCA by reef
```{r}
ind_coords.full

ind_coords.full <- ind_coords.full %>% dplyr::mutate(Site = factor(Site, levels = c("Osprey", "Bougainville", "Moore", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick","Wreck")))


# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

PCA <- ind_coords.full %>% 
  ggplot(aes(Axis1, Axis2, fill = Site)) +
  geom_point(alpha = 1, shape = 21, size = 2) + 
  scale_fill_brewer(palette = "Spectral") +
    # custom labels
  labs(x = xlab, y = ylab) +
  theme_PCA

PCA

acro_coords <- ind_coords.full

save(acro_coords, file = "acro_coords.RData")
```


##4.6 Plot PCA by latitude
```{r}
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

PCA.lat <- ind_coords.full %>% 
  ggplot(aes(Axis1, Axis2, fill = Lat)) +
  geom_point(alpha = 1, shape = 21, size = 2) + 
  scale_fill_distiller(palette = "Spectral", trans = "reverse") +
  labs(x = xlab, y = ylab) +
  theme_PCA

PCA.lat
```


##4.7 Plot PCA by longitude
```{r}
PCA.lon <- ind_coords.full %>% 
  ggplot(aes(Axis1, Axis2, fill = Long)) +
  geom_point(alpha = 1, shape = 21, size = 2) + 
  scale_fill_distiller(palette = "BrBG", trans = "reverse") +
    # custom labels
  labs(x = xlab, y = ylab) +
  theme_PCA

PCA.lon
```




#6. Admixture analysis
```{r}
library(LEA)
## Here we are moving on to the LEA package in order to estimate K clusters and create admixture plots 
```

##6.1 Convert to geno format
```{r}
gl2geno(gl3, outfile = "gl_geno", outpath = getwd(), verbose = NULL)
```

##6.2 Plot variance
```{r, include = FALSE}
pc = pca("gl_geno.lfmm", scale = TRUE)
tw = tracy.widom(pc)

#Plots the percentage of variance explained by eah component. You can look at the 'elbow' to inform the number of genetic clusers. It looks like ours is between 3-5.
plot(tw$percentage, pch = 19, col = "darkblue", cex = .8)
```


## 6.3 SNMF algorithm
```{r, include = FALSE}
snmf1 = snmf("gl_geno.geno",
        K = 1:10,    #number of K ancestral populations to run
        entropy = TRUE,     #calculate cross-entropy
        repetitions = 10,    #10 repetitions for each K
        project = "new")

plot(snmf1, col = "blue", cex = 1.5, pch = 19)
```

##6.4 Extract cross-entropy
```{r}
# Extract the cross-entropy of all runs where K = 3
ce = cross.entropy(snmf1, K = 3)
ce

lowest.ce = which.min(ce)
lowest.ce
```

##6.5 Create qmatrix 
```{r}
qmatrix = as.data.frame(Q(snmf1, K = 3, run = lowest.ce))
head(qmatrix)

# Label column names of qmatrix
ncol(qmatrix)
cluster_names = c()
for (i in 1:ncol(qmatrix)){
  cluster_names[i] = paste("Cluster", i)
}
cluster_names
colnames(qmatrix) = cluster_names
head(qmatrix)


# Add individual IDs
qmatrix$Ind = indNames(acro_gen)

#Add site IDs
qmatrix$Site = acro_gen$pop
head(qmatrix)

#Convert dataframe to long format
qlong = melt(qmatrix, id.vars=c("Ind","Site"))
head(qlong)

save("qmatrix", file = "qmatrix.Rdata")
```


## 6.6 Admixture sorted by q-value
```{r}
library(ggplot2)
library(forcats)
library(ggthemes)
library(patchwork)

pal = colorRampPalette(c("#E37769", "#EAB63E","#355952"))
cols = pal(length(unique(qlong$variable)))

# Reorder factor levels within groups
qlong <- qlong %>%
  group_by(variable) %>%
  mutate(Ind = fct_reorder(Ind, value, .desc = TRUE)) %>%
  arrange(variable, Ind) %>% 
  filter(Ind != c("A ten 271", "A ten 289"))

qlong2 <- qlong %>% mutate(variable = recode(variable,
                           "Cluster 1" = "AHCL2",
                           "Cluster 2" = "AHCL1",
                           "Cluster 3" = "AHCL3")) %>% 
  mutate(variable = factor(variable, levels = c("AHCL1", "AHCL2", "AHCL3")))

```


```{r}
#reorder by q-value
library(ggplot2)
library(forcats)
library(ggthemes)
library(patchwork)

pal = colorRampPalette(c("#E3856B", "#EEC95C","#80C4B7"))
cols = pal(length(unique(qlong$variable)))

# Reorder factor levels within groups
qlong2 <- qlong2 %>%
  group_by(variable) %>%
  mutate(Ind = fct_reorder(Ind, value, .desc = TRUE)) %>%
  arrange(variable, Ind)
```


##6.7 Admixture faceted by reef
```{r}
library(forcats)

# Relevel the "Site" variable
qlong2$Site <- forcats::fct_relevel(qlong2$Site, c("Osprey", "Bougainville", "Moore", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick","Wreck"))

# Relevel the "Ind" variable by the "Site" variable in the data frame
qlong2$Ind <- factor(qlong2$Ind, levels = unique(qlong2$Ind[order(qlong2$Site)]))


admix.fig <- ggplot(qlong2, aes(x = Ind, y = value, fill = variable)) +
  geom_col(color = "gray", linewidth = 0.1, width = 1) +
  facet_grid(~ Site, scales = "free_x", space = "free_x", switch = "x") +
  theme_minimal() +
  scale_fill_manual(values = cols) +
  labs(x = "Individual", y = "Admixture Proportion") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme(
    panel.spacing.x = unit(0.1, "lines"),
    panel.grid = element_blank(),
    axis.text.x = element_blank(),
    strip.text = element_text(colour = "black", size = 6),
    panel.background = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 8),
    plot.margin = margin(t = 30, r = 10, b = 30, l = 10, unit = "mm"),
    panel.border = element_rect(color = "black", fill = NA, size = 0.5)
  ) 

admix.fig
```


#7. Cryptic species analysis

##7.1 Grouping by 3 clusters
```{r}
#assigning each sample as one of the 3 clusters, whichever is > 50% of sample proportion 

qlongC1 <- qlong2 %>% dplyr::filter(variable == "AHCL1") %>% 
  filter(value > 0.35)

qlongC2 <- qlong2 %>% dplyr::filter(variable == "AHCL2") %>% 
  filter(value > 0.35)

qlongC3 <- qlong2 %>% dplyr::filter(variable == "AHCL3") %>% 
  filter(value > 0.35)

qlongC1C2 <- qlongC1 %>% full_join(qlongC2)

qlong.full.acro <- qlongC1C2 %>% full_join(qlongC3) %>% arrange(Ind) %>% 
  rename(cluster = variable)

save(qlong.full.acro, file = "qlong.full.acro.RData")
```



```{r}
#proportion of cluster per population 
qlong2

pop.cluster <- qlong2 %>% group_by(Site, variable) %>% summarise(mean = mean(value))

pop.cluster <- pop.cluster %>% spread(key = "variable", value = "mean") %>% 
  rename(Reef = Site)

save(pop.cluster, file = "pop.cluster.RData")
```



##7.2 K-cluster PCA
```{r}
ind_coords.admix <- ind_coords.full %>% 
  left_join(qlong.full.acro)

PCA.admix <- ind_coords.admix %>% 
  ggplot(aes(Axis1, Axis2, fill = cluster)) +
  geom_point(alpha = 1, shape = 21, size = 2) + 
    scale_fill_manual(values = cols)+
  labs(x = xlab, y = ylab) +
  theme_PCA

PCA.admix

ind_coords.admix %>% 
  group_by(cluster) %>% 
  summarise(n = n())

ind_coords.admix
```
**Important to note that admixture can inform on more recent hybridisation, BUT is not suitable to infer demographic history. You will need to run demographic models to inform of long term demographic history (e.g. Dstatistics, demographic modelling). Notes taken from https://speciationgenomics.github.io/ADMIXTURE/**



**From Rellstab: Neutral population genetic structure is defined as allele frequency differences among populations that have arisen due to neutral processes such as genetic drift, gene flow and mutation. The patterns of differences in allele frequencies among populations are the background against which loci contributing to local adaptation — a non-neutral process — are assessed in EAA. ** 
- Neutral processes affect all loci across a genome,whereas non-neutral processes affect only a subset of loci. 
- Corrections for neutral genetic structure are important in EAA, because neutral genetic structure can mimic patterns expected under non-neutral processes



